specs/opcode/1dSAR.md

```diff
@@ -0,0 +1,177 @@
+# SAR opcode
+
+## Procedure
+
+The `SAR` opcode shifts the bits towards the least significant one. The bits moved before the first one are discarded, the new bits are set to 0 if the previous most significant bit was 0, otherwise the new bits are set to 1.
+
+### EVM behavior
+
+Pop two EVM words `shift` and `a` from the stack, and push `b` to the stack, where `b` is calculated as `b = a >> shift`.
+
+Both `a` and `b` are considered as `signed` 256-bit values.
+
+### Circuit behavior
+
+To prove the `SAR` opcode, we first get the stack word `shift`, `a` and `b` to construct a gadget that proves `a >> shift == b`.
+
+The `shift` is an `unsigned` value, but both `a` and `b` are `signed` values. A `signed` value is either negative or non-negative. The value is negative if the highest bit is 1, otherwise it is non-negative.
+
+As usual, we use 32 cells to represent word `a` and `b`, where each cell holds an 8-bit value. Then split each word into four 64-bit limbs denoted by `a64s[idx]` and `b64s[idx]` where idx in `(0, 1, 2, 3)`.
+
+We put the lower `n` bits of a limb into the `lo` array, and put the higher `64 - n` bits into the `hi` array, where `n = shift % 64`. During the right shift operation, the `lo` array will move to higher bits of the result, and the `hi` array will move to next lower bits.
+
+#### Special case (shift < 64)
+
+The following figure illustrates how `SAR` opcode works under the case of `shift < 64`.
+```
++-------------------------------+-------------------------------+-----
+|a0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11| 12| 13| 14| 15| ...
++-------------------------------+-------------------------------+-----
+|             a64s[0]           |             a64s[1]           | ...
++------------+------------------+------------+------------------+-----
+| a64s_lo[0] |    a64s_hi[0]    | a64s_lo[1] |    a64s_hi[1]    | ...
++------------+------------------+------------+------------------+-----
+             |            b64s[0]            |            b64s[1]
+             +-------------------------------+------------------------
+```
+
+First we could define below constants for calculating `b64s`.
+```
+MAX_U64 = 2**64 - 1
+is_neg = is_neg(a)
+shf_div64 = shift // 64
+shf_mod64 = shift % 64
+p_lo = 1 << shf_mod64
+p_hi = 1 << (64 - shf_mod64)
+# The top new bits are set to 1 if `a` is negative, otherwise set to 0.
+p_top = is_neg * (MAX_U64 - p_hi + 1))
+a64s = word_to_64s(a)
+a64s_lo[idx] = a64s[idx] % p_lo
+a64s_hi[idx] = a64s[idx] / p_lo
+```
+
+Under this special case of `shift < 64`, `b64s` could be calculated as:
```

Comment:
This is the same case as described above. Didn't you mean `shift > 64`??

Reply:
No. It is the same case as above. Just list about how to calculate `b64s` array when `shift < 64` (with upper constants) as:
```
b64s[0] = a64s_hi[0] + a64s_lo[1] * p_hi
```
If `64 < shift < 128`, it should be (and as follows):
```
b64s[0] = a64s_hi[1] + a64s_lo[2] * p_hi
```